### DISCUSSION TITLE
Organize pvlib.bifacial?

### DISCUSSION THREAD

**Discussion by cwhanse** [pvlib/pvlib-python#1318] [Ideas]
How should we organize `pvlib.bifacial` to accommodate more bifacial irradiance models? Options appear to be:

1. leave as is. Probably OK in the short term but at some point it will become a very large file. 
2. supplement pvlib.bifacial with additional modules pvlib.bifacial_2d.py, pvlib.bifacial_ray_tracing.py with "_2d.py" containing models using 2D view factor geometries, etc.
3. turn bifacial into a folder containing several modules. Modules could be named for the source of the model or for shared utility functions (e.g. pvlib.bifacial.pvfactors, pvlib.bifacial.marion, pvlib.bifacial.util)

**Current structure**

Currently, `pvlib.bifacial` contains one function `pvfactors_timeseries`, which is a convenience wrapper for using pvfactors from pvlib `pvlib.bifacial`'s module description is "The ``bifacial`` module contains functions for modeling back surface plane-of-array irradiance under various conditions."

I like the simplicity of `pvlib.bifacial.pvfactors_timeseries` etc. This is Option #1 above. But I am concerned about the likely large file length.

**Bifacial modeling approaches**

There are two types of bifacial irradiance models:

1. View factor models, which can be either 2D or 3D. 
  - Marion’s bifacial_vf is a 2D, infinite rows (sheds) geometry view factor model. Bifacial_vf builds up rear surface irradiance by explicitly calculating its components (e.g., ground reflection, sky diffuse, reflection from adjacent rows).
  - Pvfactors is also a 2D infinite rows geometry model. It differs from bifacial_vf primarily in its numerical method (solving for components of rear irradiance through radiosity algebra).
  - The only open source 3D model I am aware of is an unfinished Sandia model which computes irradiance cell by cell. Matlab code is available. There is at least one commercial package which use 3D geometries and view factors.

2. Ray tracing models, such as bifacial_radiance.

I think implementing (rather than wrapping) 2D view factor bifacial models in pvlib makes sense. PR #717, #914 and #1187 are attempts to add 2D view factor models. I don’t think implementing anything more than a wrapper for ray tracing models makes sense for pvlib.

**Comment by kandersolar** [comment]
I favor Option 3.  If "I like the simplicity of `pvlib.bifacial.pvfactors_timeseries`" is referring to keeping it simple to import the function, e.g.:
- good: `from pvlib.bifacial import pvfactors_timeseries`
- bad: `from pvlib.bifacial.pvfactors import pvfactors_timeseries`

that can be achieved via `pvlib/bifacial/__init__.py` like how `pvlib.iotools` and `pvlib.spectrum` do it.  In fact we might also consider doing that for `irradiance.py` and others.  Edit: this is what @mikofski proposed here: https://github.com/pvlib/pvlib-python/pull/1187#issuecomment-796441452

> I think implementing (rather than wrapping) 2D view factor bifacial models in pvlib makes sense.

Does this include models with existing public implementations ([pvfactors](https://github.com/SunPower/pvfactors), [bifacial_vf](https://github.com/NREL/bifacialvf))?  I don't think I see much value in replacing a pvfactors wrapper with a re-implementation of the algorithm. 

> I don’t think implementing anything more than a wrapper for ray tracing models makes sense for pvlib.

+1 to not implementing ray tracers in pvlib.  For something as complex as bifacial_radiance I'm not sure even a wrapper is a good idea.

**Comment by cwhanse** [comment]
> Does this include models with existing public implementations (pvfactors, bifacial_vf)?

I didn't mean to imply re-implementing existing code like bifacial_vf in pvlib. I meant implementing 2D view factor models in general, because a) they are most popular and b) I think they can be maintained in pvlib. The existence of bifacial_vf doesn't mean that a pvlib-native 2D v.f. model isn't valuable, but we can discuss critiques of bifacial_vf elsewhere.

> I don't think I see much value in replacing a pvfactors wrapper with a re-implementation of the algorithm.

neither do I.

**Comment by mikofski** [comment]
I like making a folder and adapting the API in the dunder file as Kevin suggested

**Comment by kandersolar** [comment]
> The only open source 3D model I am aware of is an unfinished Sandia model which computes irradiance cell by cell. Matlab code is available. There is at least one commercial package which use 3D geometries and view factors.

@cwhanse I assume the model in question is this one ([code](https://pvpmc.sandia.gov/pv-research/bifacial-pv-project/bifacial-pv-performance-models/ray-tracing-models-for-backside-irradiance/view-factor-models/sandia-view-factor-model-implementation/), [reference](https://ieeexplore.ieee.org/abstract/document/8366707)).  It is now, or will it ever be, considered "finished" enough to be suitable for porting to pvlib-python?

Also curious if you know of any other published 3D vf models we might consider implementing in pvlib-python.  I have been daydreaming about numba-accelerated VF models but have not kept up with the literature.

**Reply by mikofski** [reply]
FYI: Annie Russell at U Ottawa said she had a 3d view factor model. Also I've found some blender/unity based 3d view factor models in the literature, mainly for bipv

**Reply by kandersolar** [reply]
DUET, right?  I briefly chatted with her about somehow getting it into pvlib but it's not clear that the university will ever allow the source code to be released (they may keep it proprietary) and it seems that the model's public documentation is not sufficient to recreate it from scratch.  Too bad because it does seem like a cool model.

**Reply by cwhanse** [reply]
@mikofski yes, that's the one. It's fast enough for a single row of modules; for an array of rows, might take a while. Using a GPU accelerates the calculation significantly. The "unfinished" aspect is mainly the coordinate system and interface for defining the physical 3D system.

Edit: the Sandia 3D Matlab code does not account for shading by array elements of other array surfaces nor for blocking of the view to the sky by array objects. It only accounts for shading of the ground by array objects.
 
I talked to the author of the DUET poster and got the same response when I asked about the availability of the code.

**Comment by mikofski** [comment]
I would ❤️  to see less new code for this and more of an effort to reuse [blender](https://www.blender.org/), [unity](https://unity.com/),  [lightstudio](https://lightstudio.com/), or any other tool that is already designed to calculate diffuse light on oblique surfaces in complex multi-surface 3D images.
